python/lsst/daf/butler/core/connectionString.py

```diff
@@ -0,0 +1,96 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("DB_AUTH_ENVVAR", "DB_AUTH_PATH", "ConnectionStringFactory")
+
+from sqlalchemy.engine import url
+from lsst.daf.butler.core.dbAuth import DbAuth
+from lsst.daf.butler.core.registry import RegistryConfig
+
+DB_AUTH_ENVVAR = "LSST_DB_AUTH"
+"""Default name of the environmental variable that will be used to locate DB
+credentials configuration file. """
+
+DB_AUTH_PATH = "~/.lsst/db-auth.yaml"
+"""Default path at which it is expected that DB credentials are found."""
+
+
+class ConnectionStringFactory:
+    """Factory for `sqlalchemy.engine.url.URL` instances.
+
+    The factory constructs a connection string URL object by parsing the
+    connection string, the 'db' key in the registry configuration.
+    Username, password, host, port or database can be specified as keys in the
+    config explicitly. If username or password are missing a matching DB is
+    found in the credentials file pointed to by `DB_AUTH_ENVVAR` or
+    `DB_AUTH_PATH` values.
+    """
+
+    keys = ('username', 'password', 'host', 'port', 'database')
+
+    def fromConfig(self, registryConfig):
```

Comment:
I think this should be a class method. `self` is not used other than to access the `keys`.

Reply:
Yes, I should have remembered to review that myself before committing. 

You are fine with the `getDialect`, `getRegistryClass` and `Registry.connectionString` being static, class and property? And the `any` test in `Butler.makeRepo`?